src/main/java/org/mastodon/util/TreeUtils.java

```diff
@@ -0,0 +1,99 @@
+package org.mastodon.util;
+
+import java.util.Collection;
+
+import org.mastodon.collection.RefCollections;
+import org.mastodon.collection.RefList;
+import org.mastodon.collection.RefSet;
+import org.mastodon.graph.Edge;
+import org.mastodon.graph.Graph;
+import org.mastodon.graph.Vertex;
+import org.mastodon.graph.algorithm.traversal.InverseDepthFirstIterator;
+
+public class TreeUtils
+{
+
+	private TreeUtils()
+	{
+		// prevent instantiation of utility class
+	}
+
+	/**
+	 * This method finds a subset of the given {@code selectedVertices} that
+	 * contains only the roots of the subtrees that are selected. The order
+	 * of the returned list follows the order of the {@code roots} list and
+	 * the order of the outgoing edges of the graphs vertices.
+	 * <p>
+	 * Example:
+	 * <pre>
+	 *   A                B
+	 *  / \             /   \
+	 * a1  a2          b1    b2
+	 *     |          / \   / \
+	 *     a3        b3 b4 b5 b6
+	 *    / \
+	 *   a4 a5
+	 * </pre>
+	 *
+	 * If {@code selectedVertices} contains: {@code {a2, a4, a5, b1, b3, b4, b6}},
+	 * then the returned list will be: {@code [a2, b1, b6]}.
+	 */
+	public static < V extends Vertex< ? > > RefList< V > findSelectedSubtreeRoots(
+			final Graph< V, ? > graph,
+			final RefList< V > roots,
+			final RefSet< V > selectedVertices )
+	{
+		final RefList< V > selectedSubtreeRoots = RefCollections.createRefList( graph.vertices() );
+		final RefSet< V > visitedNodes = RefCollections.createRefSet( graph.vertices() );
+
+		for ( final V realRoot : roots )
+			for ( final DepthFirstIteration.Step< V > step : DepthFirstIteration.forRoot( graph, realRoot ) )
+				if ( ensureNoLoop( step, visitedNodes ) && !step.isSecondVisit() ) {
+
+					final V node = step.node();
+					if ( selectedVertices.contains( node ) )
+					{
+						selectedSubtreeRoots.add( node );
+						step.truncate(); // don't visit the child nodes
+					}
+				}
+
+		return selectedSubtreeRoots;
+	}
+
+	private static < V extends Vertex< ? > > boolean ensureNoLoop( DepthFirstIteration.Step< V > step, RefSet< V > visitedNodes )
+	{
+		if ( !step.isFirstVisit() )
+			return true;
+
+		boolean isLoop = !visitedNodes.add( step.node() ); // The depth first iteration enters a node for the second time. -> there's a loop.
+		if ( isLoop )
+			step.truncate(); // Break the loop by not visiting the child nodes.
+
+		return !isLoop;
+	}
+
```

Comment:
```suggestion
/**
	 * This method finds the root nodes of trees relating to the
	 * given {@code selectedVertices}.
	 * <p>
	 * Example:
	 * <pre>
	 *   A                B
	 *  / \             /   \
	 * a1  a2          b1    b2
	 *     |          / \   / \
	 *     a3        b3 b4 b5 b6
	 *    / \
	 *   a4 a5
	 * </pre>
	 *
	 * If {@code selectedVertices} contains: {@code {a2, a4, a5, b1, b3, b4, b6}},
	 * then the returned list will be: {@code [B, A]}. The order of the returned
	 * root nodes relates to the backward order of how the root nodes were added
	 * to the {@code graph}.
	 * <p>
	 * If {@code selectedVertices} contains: {@code {a2, a4}},
	 * then the returned list will be: {@code [A]}.
	 */
```

May I suggest this Javadoc comment to be added to this very valuable method? I think, it would be even more valuable with that comment, since it was a bit unclear to me what a "real root" is.

Reply:
The method return a list. So it would be very confusing to comment on the order of the returned list.

Reply:
I agree. Then maybe leave out the sentence re order, e.g. like this:

/**
	 * This method finds the root nodes of trees relating to the
	 * given {@code selectedVertices}.
	 * <p>
	 * Example:
	 * <pre>
	 *   A                B
	 *  / \             /   \
	 * a1  a2          b1    b2
	 *     |          / \   / \
	 *     a3        b3 b4 b5 b6
	 *    / \
	 *   a4 a5
	 * </pre>
	 *
	 * If {@code selectedVertices} contains: {@code {a2, a4, a5, b1, b3, b4, b6}},
	 * then the returned set will contain: {@code [A, B]}.
	 * <p>
	 * If {@code selectedVertices} contains: {@code {a2, a4}},
	 * then the returned set will contain: {@code [A]}.
	 */

Reply:
I updated the comment and pushed it just now.